scripts/mbedtls_dev/c_build_helper.py

```diff
@@ -128,18 +131,32 @@ def get_c_expression_values(
         )
         c_file.close()
         cc = os.getenv('CC', 'cc')
-        cc_is_msvc = os.path.split(cc)[1].lower() in ('cl', 'cl.exe')
         cmd = [cc]
+
+        global _cc_is_msvc #pylint: disable=global-statement,invalid-name
+        if _cc_is_msvc is None:
+            proc = subprocess.Popen(cmd,
+                                    stdout=subprocess.DEVNULL,
+                                    stderr=subprocess.PIPE,
+                                    universal_newlines=True)
+            _cc_is_msvc = 'Microsoft (R) C/C++ Optimizing Compiler' in \
+                         proc.communicate()[1]
+
         cmd += ['-I' + dir for dir in include_path]
-        # MSVC has deprecated using -o to specify the output file.
-        output_opt = '-Fe' if cc_is_msvc else '-o'
-        cmd += [output_opt + exe_name]
+        if _cc_is_msvc:
+            # MSVC has deprecated using -o to specify the output file,
+            # and produces an object file in the working directory by default.
+            cmd += ['-Fe' + exe_name, '-Fo' + obj_name]
+        else:
+            cmd += ['-o' + exe_name]
         subprocess.check_call(cmd + [c_name])
         if keep_c:
             sys.stderr.write('List of {} tests kept at {}\n'
                              .format(caller, c_name))
         else:
             os.remove(c_name)
+        if _cc_is_msvc:
+            os.remove(obj_name)
```

Comment:
The `.obj` file should be removed in the `finally` cleanup, not just on the success path.